src/crop-manager.sol

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+interface VatLike {
+    function live() external view returns (uint256);
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function dai(address) external view returns (uint256);
+    function fork(bytes32, address, address, int256, int256) external;
+    function frob(bytes32, address, address, address, int256, int256) external;
+    function flux(bytes32, address, address, uint256) external;
+    function hope(address) external;
+    function nope(address) external;
+}
+
+interface CropLike {
+    function gem() external view returns (address);
+    function ilk() external view returns (bytes32);
+    function join(address, address, uint256) external;
+    function exit(address, address, uint256) external;
+    function tack(address, address, uint256) external;
+    function flee(address) external;
+}
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transferFrom(address, address, uint256) external;
+}
+
+contract UrnProxy {
+    address immutable public vat;
+    address immutable public usr;
+    address public owner;
+
+    constructor(address vat_, address usr_) public {
+        owner = msg.sender;
+        vat = vat_;
+        usr = usr_;
+        VatLike(vat_).hope(msg.sender);
+    }
+
+    function migrate(address newOwner) external {
+        require(msg.sender == owner, "UrnProxy/not-owner");
+        VatLike(vat).hope(newOwner);
+        VatLike(vat).nope(owner);
+        owner = newOwner;
+    }
+}
+
+contract CropJoinManager {
+    mapping (address => uint256) public wards;
+    mapping (address => address) public proxy;  // UrnProxy per user
```

Comment:
What's the motivation for putting these explicitly in the `proxy` contract? I can see at least one advantage (`proxy` lookups are able to skip an `SLOAD` when e.g. done from the Clipper), so maybe that's enough of a benefit over a completely generic proxy, but wanted to see what else I might be missing.

Reply:
Conclusion: leave it.